### PR TITLE
Initialize member data in PerigeeTrajectoryError

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

This should fix the UBSAN warnings.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by jpata** [comment]
+reconstruction
- technical

File: TrackingTools/TrajectoryParametrization/interface/PerigeeTrajectoryError.h [hunk 1]
```diff
@@ -70,7 +70,7 @@ class PerigeeTrajectoryError {
 private:
   AlgebraicSymMatrix55 thePerigeeError;
   mutable AlgebraicSymMatrix55 thePerigeeWeight;
-  mutable int inverseError;
-  mutable bool weightIsAvailable;
+  mutable int inverseError = 0;
+  mutable bool weightIsAvailable = 0;
```
Comment (perrotta) [inline TrackingTools/TrajectoryParametrization/interface/PerigeeTrajectoryError.h:74]:
```suggestion
  mutable bool weightIsAvailable = false;
```

Comment (Dr15Jones) [reply TrackingTools/TrajectoryParametrization/interface/PerigeeTrajectoryError.h]: Done. Thanks!

**Comment by VinInn** [comment]
Do not really understand this one.
Is this not a false positive?
and Why initialized to zero?

**Comment by Dr15Jones** [comment]
This stems from the UBSAN IB

> Is this not a false positive?

https://cmssdt.cern.ch/SDT/jenkins-artifacts/ubsan_logs/CMSSW_12_3_X_2022-02-05-1100/logs/5b/5b0b48448feee12e3fd3166c553fb58f/log

e.g.
```
1329.0/step3:TrackingTools/TrajectoryParametrization/interface/PerigeeTrajectoryError.h:14:7: runtime error: load of value 173, which is not a valid value for type 'bool'
```
Based on all the reports I've looked at, it appears the UBSAN only detects when bools are not properly initialized and doesn't do so for any other builtin types. The few cases I've more closely looked at happen when either the copy constructor or `operator=` are being called with the _right hand side_ being a instance of the object which was default constructed. I assume that is what is happening here.

> and Why initialized to zero?

I've found that people who do not know that C++ doesn't initialize builtin type member data assume it will be initialized to 0 (so I was just following that chain of thought).
You are free to suggest an alternative value.

**Comment by VinInn** [comment]
>The few cases I've more closely looked at happen when either the copy constructor or operator= are being called with the right hand side being a instance of the object which was default constructed. I assume that is what is happening here.

This is what I would call a "false positive"  and I object to fix correct code only because a tool does not understand the logic.

**Comment by jpata** [comment]
So how do we want to proceed here?

Do tracking experts confirm the behavior is intended (and UBSAN should be suppressed here), or should this be investigated further?

Does UBSAN give some more information about where the uninitialized use is happening?

**Comment by smuzaffar** [comment]
@Dr15Jones , I will enable `-g` option for UBSAN IBs to too if that gives more detail.

**Comment by smuzaffar** [comment]
https://github.com/cms-sw/cmsdist/pull/7608 should build UBSAN in debug mode  now

**Comment by Dr15Jones** [comment]
I did a bit of poking about.

An invalid PerigeeTrajectoryError is created in this constructor 
https://github.com/cms-sw/cmssw/blob/e79891995244caf06e9f3f11c86242a793af4222/TrackingTools/TrajectoryState/interface/TrajectoryStateClosestToPoint.h#L31-L41

then a call to 
https://github.com/cms-sw/cmssw/blob/e79891995244caf06e9f3f11c86242a793af4222/RecoVertex/VertexTools/interface/PerigeeLinearizedTrackState.h#L237-L245

can be made where `thePredState` has the uninitialized member data (since call to `computeJacobians()` which fills `thePredState` ultimately traced back to here
https://github.com/cms-sw/cmssw/blob/e79891995244caf06e9f3f11c86242a793af4222/TrackingTools/TrajectoryState/src/TrajectoryStateClosestToPoint.cc#L6-L15
)

which mean the returned info from the call to `weightMatrix` will be unknown.

This does seem to mean that `inverseError` should probably be initialized to `1` to match what was done above.

**Comment by jpata** [comment]
thanks, let see what the tracking experts suggest here.
also tagging @cms-sw/tracking-pog-l2

**Comment by jpata** [comment]
So should we instead use `CMS_SA_ALLOW` here?

**Comment by makortel** [comment]
> So should we instead use `CMS_SA_ALLOW` here?

`CMS_SA_ALLOW` is recognized only by static analyzer, so won't do anything for UBSAN.

**Comment by jpata** [comment]
@VinInn would you have some time to suggest a fix here? I was not able to dig up if just replacing by 1 is fine here or not, sorry.

**Comment by VinInn** [comment]
I would just leave it as it was...

**Comment by VinInn** [comment]
or add "default" copy and move constructors (and assign operators)

**Comment by VinInn** [comment]
or fix the logic of the user of ```TrajectoryStateClosestToPoint```
if error is not available it should not be asked for...

**Comment by Dr15Jones** [comment]
> or fix the logic of the user of TrajectoryStateClosestToPoint if error is not available it should not be asked for...

I just happened to spot that problem. There could be addition cases where the uninitialized data is read and there could be more in the future as people change/add code.

**Comment by VinInn** [comment]
How the class shall be used is clear: it has a method to ask if the error is available.
If people misuse it they get UB. By design.

**Comment by jpata** [comment]
It looks to me based on the above discussion that
1) some action is needed to either remove or prevent the possibility of the UB (just thinking out loud, could it be better to abort the job rather than have undefined behaviour?)
2) it's not trivial how to do it, and the tracking-pog should have a say in it
so it would make sense perhaps to close this PR, and open an issue.

Any objections?

**Comment by VinInn** [comment]
> On 17 Feb, 2022, at 5:38 PM, Joosep Pata ***@***.***> wrote:
> 
> 	• some action is needed to either remove or prevent the possibility of the UB (just thinking out loud, could it be better to abort the job rather than have undefined behaviour?)

any uninitialized bool allocated on the heap will upset UBSAN. (not other type though, uint8_t is ok!)
It is a false positive. nothing to fix (but the tool).

and there is nothing to abort. the code is correct, there is no evidence that anybody is asking for the error when none is provided.

v.

**Comment by jpata** [comment]
-reconstruction

As I understand, the fix proposed here by Chris is not universally accepted, so the tracking POG may be in a position decide here if/how this is to be fixed. I found an older issue for this https://github.com/cms-sw/cmssw/issues/35036#issuecomment-1044139169, where I updated the status. There are a few more cases in tracking packages.

**Comment by Dr15Jones** [comment]
under protest I will close this

**Pull request closed without merging**